### DISCUSSION TITLE
Possible Bug in Class Property Inference?

### DISCUSSION THREAD

**Discussion by shininglove** [microsoft/pyright#5229] [General]
Found a bug in my code because a None wasn't being handled and when I discovered why, it was because a property was returning back the wrong type.

![user_types](https://github.com/microsoft/pyright/assets/87951084/5acdb364-416a-4f90-8171-4b0307e6db46)

Even though when I go to the library and check the type its fine there:
![content_type_i](https://github.com/microsoft/pyright/assets/87951084/d6fe5ea3-3c4d-43da-85ee-71d81c37538e)
^ This is located in the env/lib/python3.x/site-packages/requests/models.py

Is this expected behavior for pyright?

**Comment by erictraut** [comment]
Please provide a self-contained, minimal code sample (in text form) that demonstrates the issue. Diagnosing issues from partial screen shots isn't really feasible.

**Comment by shininglove** [comment]
Sorry about that!
Here is an isolated example of it:

```python
import json
import requests

example_info = requests.get("https://example.com/posts")

# content returns only type bytes 
# instead of the three types it could be
data = json.loads(example_info.content)

print(data)
```

**Reply by erictraut** [reply]
Thanks for the sample. I'm not sure what behavior you think is incorrect.

The call to `requests.get` returns a value of type `Response` according to the type stub `requests/requests/models.pyi`. The `Response` class has a property called `content` that is declared to return `bytes`. This is declared in the same typestub file. Here's the relevant portion:

```python
class Response:
    @property
    def content(self) -> bytes: ...
```

The call to `json.loads` is declared to return `Any` in the type stub `json/__init__.pyi`. Here's the declaration:

```python
def loads(
    s: str | bytes | bytearray,
    *,
    cls: type[JSONDecoder] | None = None,
    object_hook: Callable[[dict[Any, Any]], Any] | None = None,
    parse_float: Callable[[str], Any] | None = None,
    parse_int: Callable[[str], Any] | None = None,
    parse_constant: Callable[[str], Any] | None = None,
    object_pairs_hook: Callable[[list[tuple[Any, Any]]], Any] | None = None,
    **kwds: Any,
) -> Any: ...
```

All of these types definitions come directly from the typeshed stubs. If you think some of these type definitions are incorrect, you can file bugs (or PRs) in the [typeshed project](https://github.com/python/typeshed). These type stubs are used by all of the major Python type checkers and language servers including pyright.

**Reply by hmc-cs-mdrissi** [reply]
Hmm reading source code for requests

```python
@property
def content(self) -> bytes  | None: ...
```

looks correct. If request follows to get a response (network down) it returns None. So this is a change to make in typeshed. It may be hit for number of repos as I suspect most people aren't doing assert response.content is not None, "Failed request." I'd recommend a pr and see the mypy primer output/typeshed maintainers can make a decision.

**Reply by shininglove** [reply]
Thank you, I wasn't sure if this was a typeshed or pyright issue because it was reporting the content type correctly when I inspected the sourcecode, but wasn't when I tried consuming the function. I
